src/replay/service.zig

```diff
@@ -91,19 +96,1269 @@ fn advanceReplay(state: *ReplayState) !void {
+        };
+    }
+};
+
+pub const DuplicateConfirmedState = struct {
+    duplicate_confirmed_hash: sig.core.Hash,
+    bank_status: BankStatus,
+};
+
+pub const DeadState = struct {
+    cluster_confirmed_hash: ?ClusterConfirmedHash,
+    is_slot_duplicate: bool,
 
-    // TODO: process_duplicate_confirmed_slots
+    pub fn fromState(
+        slot: sig.core.Slot,
+        duplicate_slots_tracker: *const DuplicateSlotsTracker,
+        duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        epoch_slots_frozen_slots: *const EpochSlotsFrozenSlots,
+    ) DeadState {
+        _ = slot; // autofix
+        _ = duplicate_slots_tracker; // autofix
+        _ = duplicate_confirmed_slots; // autofix
+        _ = fork_choice; // autofix
+        _ = epoch_slots_frozen_slots; // autofix
+    }
+};
+
+pub const DuplicateState = struct {
+    duplicate_confirmed_hash: ?sig.core.Hash,
+    bank_status: BankStatus,
+
+    pub fn fromState(
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        bank_status: BankStatus,
+    ) DuplicateState {
+        // We can only skip marking duplicate if this slot has already been
+        // duplicate confirmed, any weaker confirmation levels are not sufficient
+        // to skip marking the slot as duplicate.
+        const duplicate_confirmed_hash = getDuplicateConfirmedHashFromState(
+            logger,
+            slot,
+            duplicate_confirmed_slots,
+            fork_choice,
+            bank_status.bankHash(),
+        );
+        return .{
+            .duplicate_confirmed_hash = duplicate_confirmed_hash,
+            .bank_status = bank_status,
+        };
+    }
+};
+
+pub const EpochSlotsFrozenState = struct {
+    epoch_slots_frozen_hash: sig.core.Hash,
+    duplicate_confirmed_hash: ?sig.core.Hash,
+    bank_status: BankStatus,
+    is_popular_pruned: bool,
+
+    pub fn fromState(
+        logger: sig.trace.Logger,
+        slot: sig.core.Slot,
+        epoch_slots_frozen_hash: sig.core.Hash,
+        duplicate_confirmed_slots: *const DuplicateConfirmedSlots,
+        fork_choice: *const sig.consensus.HeaviestSubtreeForkChoice,
+        bank_status: BankStatus,
+        is_popular_pruned: bool,
+    ) EpochSlotsFrozenState {
+        const duplicate_confirmed_hash = getDuplicateConfirmedHashFromState(
+            logger,
+            slot,
+            duplicate_confirmed_slots,
+            fork_choice,
+            bank_status.bankHash(),
+        );
+        return .{
+            .epoch_slots_frozen_hash = epoch_slots_frozen_hash,
+            .duplicate_confirmed_hash = duplicate_confirmed_hash,
+            .bank_status = bank_status,
+            .is_popular_pruned = is_popular_pruned,
+        };
+    }
+};
+
+fn processAncestorHashesDuplicateSlots(
```

Comment:
I know this is generally not consistently done, but can you help add a link that points to the analogous function in Agave?

Also only looking at the input params, I noticed a couple are missing when compared to Agave:

- `duplicate_slots_tracker: &mut DuplicateSlotsTracker,`
- `ancestor_hashes_replay_update_sender: &AncestorHashesReplayUpdateSender`
- `purge_repair_slot_counter: &mut PurgeRepairSlotCounter`

Is this omission intentional?

Reply:
115f85961231a55bc485639d4026ae631c828e07